Server

FlandiaYingman opened the pull request.

tonyyuyiding commented:
> @tonyyuyiding
> 
> Could you work on the email notification thing? In summary, we will have to create a service `NotificationService`, and in implementation, it can use SMTP to send emails.
> 
> Here are the information from CSE System:
> 
> > Yes, you may use the [no_reply_crs@cse.ust.hk](mailto:no_reply_crs@cse.ust.hk) as your sender address and set the SMTP server to smtp.cse.ust.hk, port 25. I have setup our system to allow mail sending from this network address (143.89.130.64) without authentication required.
> 
> So basically they allow us to send emails on our server without auth. But we still need a mechanism to test email sending locally. We'll probably need to use our own email account for testing.

No problem.

FlandiaYingman commented:
> I just added a simple email service. This is just a quick demo. 

Thanks. Did you try it out in practice. Could you trigger an email to me, you, and/or Desmond to see how it works?

> We may need to consider how to integrate it into the backend or fetch the `to` and `cc` lists automatically.

I guess that could be in the service layer. Let me think of it a bit.

---

BTW I just implemented the authentication on both the server-side and client-side. We just have to work on

- [ ] Containerization (for Deployment)
- [ ] Instructor's Form Submission (the User's one is done)

FlandiaYingman commented:
> Sure. I have just triggered an email to you. Most likely you will find it in your spam folder >.<

Oops, I didn't receive any email. Is it blocked by the mail server?

> btw I saw you pushed some `.env` files. Is it safe to do that? This repo is public. Besides, if you need the env settings for testing the emails, I can send you in some secret way.

Yep I only pushed the `.env` containing safe info. That's to track what env var that the app requires. I filled in only non-sensitive info there, like the auth `CLIENT_ID` and `TENANT_ID`. Sensitive information, including `CLIENT_SECRET`, is not pushed.

FlandiaYingman commented on packages/service/lib/notificationService.ts:
```diff
@@ -0,0 +1,92 @@
+    templateName: string,
+    context: Record<string, string>,
+  ): Promise<string> {
+    const templatePath = path.join(this.templateDir, templateName);
+    const source = await fs.readFile(templatePath, "utf-8");
```
Let's use `Bun.file` instead of Node's funny API

FlandiaYingman commented:
> * fetching enrollment data from itso

I checked with Kris. He says that this is sensitive so will be more difficult to do it. Better way is to make a converter, e.g., copy the names / email from other system (e.g. Canvas), and manually type in to our system.

> * admin page for instructors to add assignments and TAs
> * (maybe) a page for TAs to view the requests

I think they will be the next milestone to work on. Let's have this version deployed first.
